### PR TITLE
Use Gurobi_jll for binaries

### LINKED ISSUE #509: A recommendation for code revision
I suggest to revise the build.jl like this
"
const ALIASES = [
    "gurobi100",
    "gurobi95",
    "gurobi91",
    "gurobi90"
]

paths_to_try = []
"

The reason is that the original piece of code "paths_to_try = copy(ALIASES)" will result in the library handle variable like "libgurobi = gurobi100".  Such a handle variable does not include the absolute path and makes a compiled package with "Gurobi.jl" throw error of "ErrorException("could not load library "gurobi 100"".

### PR DESCRIPTION
**Pull request by odow**
See https://github.com/jump-dev/Gurobi_jll.jl

There are still a few unresolved issues:

 - [x] Tag a release of Gurobi_jll
 - [x] Check if binaries are installed on `Pkg.add`, or only if needed
 - [x] Document how to set a license if using provided binaries
 - [x] Test locally

A potential first-pass is to use the binaries only on CI (or linux, to maintain compat).

Would fix issues like 
https://discourse.julialang.org/t/how-to-use-gurobi-11-in-julia-launched-by-anaconda/111271/2

Closes #509

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -4,36 +4,37 @@ on:
...
 jobs:
   test:
     name: 'Gurobi'
-    runs-on: 'ubuntu-latest'
+    runs-on: ${{ matrix.os }}
     strategy:
       fail-fast: false
+      matrix:
+        version: ['1.6', '1']
+        os: [ubuntu-latest, macOS-latest, windows-latest]
```
Comment (odow) [inline .github/workflows/ci.yml:19]:
```suggestion
        os: [ubuntu-latest, macOS-latest, macos-14, windows-latest]
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -4,36 +4,42 @@ on:
...
-          path: ~/.julia/artifacts
-          key: ${{ runner.os }}-test-${{ env.cache-name }}-${{ hashFiles('**/Project.toml') }}
-          restore-keys: |
-            ${{ runner.os }}-test-${{ env.cache-name }}-
-            ${{ runner.os }}-test-
-            ${{ runner.os }}-
-      - shell: bash
+          version: ${{ matrix.version }}
+          arch: ${{ matrix.arch }}
+      - uses: julia-actions/cache@v1
```
Comment (odow) [inline .github/workflows/ci.yml:31]:
```suggestion
      # - uses: julia-actions/cache@v1
```

**Comment by odow** [comment]
Hmm. Something isn't right:
```
(base) oscar@Oscars-MBP /tmp % julia --project=xpr
The latest version of Julia in the `1.10` channel is 1.10.2+0.x64.apple.darwin14. You currently have `1.10.1+0.x64.apple.darwin14` installed. Run:

  juliaup update

to install Julia 1.10.2+0.x64.apple.darwin14 and update the `1.10` channel to that version.
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.1 (2024-02-13)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(xpr) pkg> st
Status `/private/tmp/xpr/Project.toml` (empty project)

(xpr) pkg> add https://github.com/jump-dev/Xpress_jll.jl
     Cloning git-repo `https://github.com/jump-dev/Xpress_jll.jl`
    Updating git-repo `https://github.com/jump-dev/Xpress_jll.jl`
   Resolving package versions...
    Updating `/private/tmp/xpr/Project.toml`
  [308bddfa] + Xpress_jll v9.3.0 `https://github.com/jump-dev/Xpress_jll.jl#master`
    Updating `/private/tmp/xpr/Manifest.toml`
  [692b3bcd] + JLLWrappers v1.5.0
  [21216c6a] + Preferences v1.4.3
  [308bddfa] + Xpress_jll v9.3.0 `https://github.com/jump-dev/Xpress_jll.jl#master`
  [56f22d72] + Artifacts
  [ade2ca70] + Dates
  [8f399da3] + Libdl
  [de0858da] + Printf
  [fa267f1f] + TOML v1.0.3
  [4ec0a83e] + Unicode
Precompiling project...
  1 dependency successfully precompiled in 2 seconds. 2 already precompiled.

julia> import Xpress_jll

(xpr) pkg> st
Status `/private/tmp/xpr/Project.toml`
  [308bddfa] Xpress_jll v9.3.0 `https://github.com/jump-dev/Xpress_jll.jl#master`

(xpr) pkg> add https://github.com/jump-dev/Gurobi_jll
     Cloning git-repo `https://github.com/jump-dev/Gurobi_jll`
ERROR: failed to clone from https://github.com/jump-dev/Gurobi_jll, error: GitError(Code:ERROR, Class:HTTP, unexpected http status code: 404)

(xpr) pkg> add https://github.com/jump-dev/Gurobi_jll.jl
    Updating git-repo `https://github.com/jump-dev/Gurobi_jll.jl`
   Resolving package versions...
    Updating `/private/tmp/xpr/Project.toml`
  [c018c7e6] + Gurobi_jll v11.0.0 `https://github.com/jump-dev/Gurobi_jll.jl#master`
    Updating `/private/tmp/xpr/Manifest.toml`
  [c018c7e6] + Gurobi_jll v11.0.0 `https://github.com/jump-dev/Gurobi_jll.jl#master`

julia> import Gurobi_jll
zsh: killed     julia --project=xpr
(base) oscar@Oscars-MBP /tmp % 
```

**Comment by odow** [comment]
The issue appears to be
```
julia> path_name = "/Users/oscar/.julia/artifacts/5a3b623cd7ac3c59c02bb5063b4db60287f33a1f/lib/libgurobi110.dylib"
"/Users/oscar/.julia/artifacts/5a3b623cd7ac3c59c02bb5063b4db60287f33a1f/lib/libgurobi110.dylib"

julia> using Libdl; dlopen(path_name, RTLD_LAZY | RTLD_DEEPBIND)
zsh: killed     julia --project=xpr
```
I wonder if it is because `libgurobi110` is already available
```Julia
julia> using Libdl

julia> dlpath("libgurobi110.dylib")
"/Library/gurobi1100/macos_universal2/lib/libgurobi110.dylib"
```

**Comment by odow** [comment]
Here's a user that this would help: https://discourse.julialang.org/t/how-to-use-gurobi-11-in-julia-launched-by-anaconda/111271/5?u=odow

**Comment by jd-foster** [comment]
I can't reproduce your issue following your steps (julia 1.10.2 (arm64 version) on M2 Pro). It seems to run fine for me but I don't have Gurobi 11 installed. I know you've thought of this already, but have you definitely ruled out Rosetta incompatibilities?

**Comment by odow** [comment]
I stopped digging into this. I was waiting for Gurobi to get back to me on whether they're okay with this general approach.

File: Project.toml [hunk 3]
```diff
@@ -4,12 +4,12 @@ repo = "https://github.com/jump-dev/Gurobi.jl"
...
 
 [deps]
-LazyArtifacts = "4af54fe1-eca0-43a8-85a7-787d91b784e3"
+Gurobi_jll = "c018c7e6-a5b0-4aea-8f80-9c1ef9991411"
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 
 [compat]
-LazyArtifacts = "<0.0.1, 1.6"
+Gurobi_jll = "=9.5.2"
```
Comment (odow) [inline Project.toml:12]:
```suggestion
Gurobi_jll = "=9.5.2, =10.0.3"
```

**Comment by odow** [comment]
Can people try this locally? I'm still getting https://github.com/jump-dev/Gurobi_jll.jl/issues/5

**Comment by simonbowly** [comment]
Tested on M3 mac and linux (intel, glibc). No issues so far!

**Comment by torressa** [comment]
Tested on an M1 and Linux, both worked!

**Comment by odow** [comment]
Yeah I've obviously borked something.

I guess one question for @simonbowly and @torressa: should have we have this enabled by default?

Xpress asked us not to: https://github.com/jump-dev/Xpress.jl?tab=readme-ov-file#use-with-xpress_jll

But I would be in favor of enabling by default. It should mean that we just start having zero installation questions because everything will magically work. We should just add documentation on how to use grbgetkey to initialize a license.

**Comment by simonbowly** [comment]
Sorry I missed this, travelling at the moment.

Yes I'd say enable it by default if that's your preference. Whatever is most familiar/easy for JuMP users. I assume it will still be possible to manually compile against a local install (say if someone needs a specific technical release not covered by Gurobi_jll)?

**Comment by odow** [comment]
> Yes I'd say enable it by default if that's your preference

Okay, great.

> I assume it will still be possible to manually build Gurobi.jl against a local install

Yes, I'll sort some documentation for this.

**Comment by odow** [comment]
I still haven't gotten to the bottom of my local crash, although I verified that things work okay on a different machine...

I propose that we take a look over the documentation, merge this, and then invite the community to test `] add Gurobi#master` before we tag a new release.

**Comment by odow** [comment]
Elliot saves the day again:

image

**Comment by odow** [comment]
@jd-foster does it work if you run `codesign --remove-signature ~/.julia/artifacts/.julia/artifacts/48494ffef82cbb1907e4dc8974431fe60a12ef06/lib/libgurobi110.dylib`?

If so, we can add this to the JLL

**Comment by odow** [comment]
See https://github.com/jump-dev/Gurobi_jll.jl/pull/13

Install with: `] add Gurobi_jll#od/fix-macos`

**Comment by jd-foster** [comment]
> does it work if you run `codesign --remove-signature`
Yes, that part also works, and installing with the `od/fix-macos` branch works too.

**Comment by odow** [comment]
Great. I'll tag a new release https://github.com/jump-dev/Gurobi_jll.jl/pull/14

**Comment by odow** [comment]
An update should do the trick:
```Julia
julia> using Gurobi

julia> Gurobi.libgurobi
"/Users/oscar/.julia/artifacts/c6002a5d5f15ef98c2821ea90c04c1ce087df37f/lib/libgurobi110.dylib"

(Gurobi) pkg> test
     Testing Gurobi
      Status `/private/var/folders/bg/dzq_hhvx1dxgy6gb5510pxj80000gn/T/jl_pAJTLT/Project.toml`
  [2e9cd046] Gurobi v1.2.3 `~/.julia/dev/Gurobi`
  [b8f27783] MathOptInterface v1.29.0
  [c018c7e6] Gurobi_jll v11.0.2
...

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Test Summary:          | Pass  Total     Time
MathOptInterface Tests | 4438   4438  4m48.2s
     Testing Gurobi tests passed 
```

**Comment by odow** [comment]
@simonbowly any comments before I merge?

**Comment by simonbowly** [comment]
No comments, looks good to me.

**Pull request merged**